plugins/optimization-detective/storage/data.php

```diff
@@ -151,14 +151,16 @@ function od_get_url_metrics_slug( array $query_vars ): string {
  *
  * @see od_verify_url_metrics_storage_hmac()
  * @see od_get_url_metrics_slug()
+ * @todo This should also include an ETag as a parameter. See <https://github.com/WordPress/performance/issues/1466>.
  *
- * @param string $slug Slug (hash of normalized query vars).
- * @param string $url  URL.
- *
+ * @param string      $slug                Slug (hash of normalized query vars).
+ * @param string      $url                 URL.
+ * @param string|null $queried_object_type Queried object type.
+ * @param int|null    $queried_object_id   Queried object ID.
  * @return string HMAC.
  */
-function od_get_url_metrics_storage_hmac( string $slug, string $url ): string {
-	$action = "store_url_metric:$slug:$url";
+function od_get_url_metrics_storage_hmac( string $slug, string $url, ?string $queried_object_type = null, ?int $queried_object_id = null ): string {
```

Comment:
Based on what we previously discussed. I think it would be great to abstract the URL representation into a custom type. For example a class like `URL_Entity` or `URL_Content` or so, which would be able to support different ways to identify content (e.g. not everything has a "queried object" or a numeric ID, some URLs need other kinds of identifiers).

Maybe have that class implement `__toString()`, which would be a perfect fit for the usage here.

Reply:
Yeah, potentially. However, that may make more sense for https://github.com/WordPress/performance/issues/1466. In particular, I'm wary to not expose too much of the internals of how the page is constructed, as these arguments have to be passed to the frontend. The exclusive purpose of this parameter here is to communicate the endpoint which object should have its cache cleaned. Since it won't always be the "queried object" in WP parlance in the case of many archives but rather could be simply the first post in the loop. In that way, there should _always_ be some object to pass here with an integer ID. Otherwise, if there is no object, then this would normally be `is_404()` case which is not optimized anyway.

Reply:
> I'm wary to not expose too much of the internals of how the page is constructed, as these arguments have to be passed to the frontend.

I don't think any of this would be sensitive - most of it is already surfacing publicly anyway, via `body` classes that WordPress Core adds.

> Since it won't always be the "queried object" in WP parlance in the case of many archives but rather could be simply the first post in the loop.

For that though, you would need to know _which_ loop. For example, if there was a custom post type archive "type", it would still be missing an identifier for which type, so that you could fetch the first post for that post type archive, not _any_ post type archive.

All of this IMO would make more sense to be abstracted away in a class. It would make passing around this data more intuitive too (e.g. in JS it would just be an object with the necessary data), since all you need to know would be to pass around one object (e.g. `urlEntity` or whatever we'd call it).

Reply:
I don't think subqueries are relevant here, are they? Why would we need to know which loop the post was in?

Most caching plugins would have no idea which URL to flush when provided the post in a subquery since there is nothing connecting the top URL query vars to the arbitrary query vars passed into the subquery. More sophisticated caching plugins _would_ be able to connect these dots if they're using surrogate keys and if they do output buffering of the page and accumulate a new key for each `the_post` action. 

Pantheon Advanced Page Cache only computes surrogate keys for the posts in the main loop: https://github.com/pantheon-systems/pantheon-advanced-page-cache/blob/e3b5552b0cb9268d9b696cb200af56cc044920d9/inc/class-emitter.php#L211-L322

LightSpeed Cache's surrogate key computation is quite similar: https://github.com/litespeedtech/lscache_wp/blob/7c707469b3c88b4f45d9955593b92f9aeaed54c3/src/tag.cls.php#L239-L313

By default it is only computing surrogate keys for posts in the main loop. It does expose [actions](https://github.com/litespeedtech/lscache_wp/blob/7c707469b3c88b4f45d9955593b92f9aeaed54c3/src/api.cls.php#L83-L91) which plugins can use to add surrogate keys for subqueries, which seems to only be used in a WooCommerce integration for its related products:https://wpdirectory.net/search/01JCKMEYKEX35RM0JVJ4TE8RV3

For our purposes here, we just need to communicate to the endpoint which object should be cleaned from the cache. Thinking about this some more, there's no need for the `queriedObject` top-level key in the URL Metric schema. We just need the endpoint to accept two parameters alongside the `url` and the `slug` (in addition to the `hmac` which validates the parameters):

* The type of object which should be cleaned from the cache.
* The ID of the object which should be cleaned from the cache.

Pantheon Advanced Page Cache does the right thing to flush term and author archives when the term or user is updated. But it seems unique in this regard. I don't see that LightSpeed Cache has [any hooks](https://github.com/litespeedtech/lscache_wp/blob/7c707469b3c88b4f45d9955593b92f9aeaed54c3/src/purge.cls.php#L51-L75) to flush caches in response to anything bot a post edit (or a comment change). This being the case, perhaps using the queried object as the basis for flushing the cache isn't ideal anyway and that we should only ever pass along the ID for the post that should have its cache cleaned. This would handle the cases for:

1. Singular post/page in which the queried object is a `post`.
1. Non-static homepage in which case there is no queried object but there is an initial `post` in the loop.
2. Author/term/date archives in which there is also an initial post in the loop.

So yeah, I think this can be further simplified so that a `cache_purge_post_id` parameter is added to the endpoint, and when the endpoint receives it, it schedules an event in WP Cron to then fire the hooks to simulate an edit to that post in order for page caches to invalidate their cache for the URL(s). In reality, this will result in more than the intended URL being purged from the cache (e.g. the post permalink but also the homepage and maybe some archive as well), but there's no way to avoid that given the lack of any WordPress APIs for page caching. It would be really nice if WordPress had a standard API for page caching plugins to be told to purge the cache for a given URL!

Reply:
I'm not entirely sure what you're referring to with "subqueries", so I only understand parts of your response.

I wouldn't tie our approach too much to what cache plugins are actually doing (particularly when there is more than they could do). That said, the approach of always clearing a specific post makes sense to me, since every URL contains at least one post in some capacity (except 404s).

What I was referring to before was the idea to pass the context around (e.g. type etc.) so that then the receiving endpoint would figure out which post to clear the cache for. I think what you're saying now is that instead OD would already determine that for the URL initially and then pass around the resulting post ID. So I think we were more or less talking about the same end goal, but with different approaches. Passing around the post ID to clear the cache for makes sense to me.

Reply:
@felixarntz Sorry, I misinterpreted what you meant by "_which_ loop". My mind went to adding additional `WP_Query` loops inside of the main loop.

I think your idea makes sense as a future enhancement to give caching plugins more say over how cache invalidation is handled. But for this first iteration, I think sending along more context would be useful eventually.

Aside: What if the query vars were sent as-is without being pre-hashed into a `slug`? Having access to these query vars would be useful for caching plugins to determine what cache should be cleared as well. They would also be useful for debugging a URL Metric. I was hesitant to expose the query vars, thinking that perhaps there could be some private data located in there, but maybe not since these query vars would all be exposed if pretty permalinks aren't enabled? Maybe I'm being overly cautious. Probably something for another PR.

Reply:
I think keeping the query vars hashed for now is best. I'm cautious of exposing more if we don't have to, so let's discuss that separately.

> I think your idea makes sense as a future enhancement to give caching plugins more say over how cache invalidation is handled. But for this first iteration, I think sending along more context would be useful eventually.

Which idea are you referring to? Based on my understanding, _your_ idea from the earlier comment is to pass around only a post ID that needs to have its cache invalidated when metrics for the URL change, instead of passing around queried object type and queried object ID. And that makes sense to me, based on the idea that the logic would determine that post ID...
* either based on the queried object ID
* or if there's no queried object ID, the first post in the current loop

Reply:
Yes, that is my idea, as that is what I have found to be most reliable to invalidate caches for caching plugins. However some caching plugins, like Advanced Pantheon Page Cache, _would_ be able to handle a more targeted cache invalidation when the queried object is a term or author. But they'll also work when invalidating the cache for a post that appears in the loop on the archive page, so I think for now what is currently implementing in this PR is a good first step.